### PR TITLE
V0.6 - sunos fixes for fs.watch scaling.

### PR DESCRIPTION
**Pull request by ilc**
Instead of using one port per watch, use one port for all the watches.

Any comments on how to clean things up a bit are appreciated.

### CODE REVIEW

File: src/unix/sunos.c [hunk 1]
```diff
@@ -89,37 +89,45 @@ void uv_loadavg(double avg[3]) {
...
-  if (port_associate(handle->fd,
-                     PORT_SOURCE_FILE,
-                     (uintptr_t) &handle->fo,
-                     FILE_ATTRIB | FILE_MODIFIED,
-                     NULL) == -1) {
-    uv__set_sys_error(handle->loop, errno);
+#define PORT_FIRED 0x69
+#define PORT_UNUSED 0x0
+#define PORT_LOADED 0x99
+#define PORT_DELETED -1
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: Defines go at the top, right below the includes.

File: src/unix/sunos.c [hunk 2]
```diff
@@ -89,37 +89,45 @@ void uv_loadavg(double avg[3]) {
...
+static void uv__fs_event_rearm(uv_loop_t *loop, uv_fs_event_t *handle) {
+  if (handle->fd != -1) {
+    if (port_associate(loop->fs_fd,
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: Prefer flat code, e.g. `if (handle->fd == -1) return;`

File: src/unix/sunos.c [hunk 3]
```diff
@@ -144,39 +152,45 @@ int uv_fs_event_init(uv_loop_t* loop,
...
 
   memset(&handle->fo, 0, sizeof handle->fo);
   handle->fo.fo_name = handle->filename;
-  uv__fs_event_rearm(handle);
+  uv__fs_event_rearm(loop, handle);
 
-  ev_io_init(&handle->event_watcher, uv__fs_event_read, portfd, EV_READ);
-  ev_io_start(loop->ev, &handle->event_watcher);
-  ev_unref(loop->ev);
+  if(first_run) {
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: Space between if and paren.

File: src/unix/sunos.c [hunk 4]
```diff
@@ -144,39 +152,45 @@ int uv_fs_event_init(uv_loop_t* loop,
...
 
   loop->counters.fs_event_init++;
 
   /* We don't support any flags yet. */
   assert(!flags);
-
-  if ((portfd = port_create()) == -1) {
-    uv__set_sys_error(loop, errno);
-    return -1;
+  if(loop->fs_fd == 0) {
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: Space between if and paren.

Comment (bnoordhuis) [inline src/unix/sunos.c]: By the way, I suggest you set `fs_fd` to -1 in `uv__loop_init()`. It's technically possible if perhaps inadvisable for `port_create()` to return 0 if stdin is closed.

**Comment by bnoordhuis** [comment]
There is some cleanup that needs to be done but in general the patch LGTM.

**Comment by ilc** [comment]
All comments are addressed.  There is no uv__loop_init, I did find uv_loop_new.  I believe that to be the right spot given that is where it is calloced.

**Comment by ilc** [comment]
Delay that.. missed something.

**Comment by ilc** [comment]
It is now clean.  There was a real regression, when I put in the fs_fd == -1 code.

File: src/unix/sunos.c [hunk 5]
```diff
@@ -144,39 +154,45 @@ int uv_fs_event_init(uv_loop_t* loop,
...
 
   loop->counters.fs_event_init++;
 
   /* We don't support any flags yet. */
   assert(!flags);
-
-  if ((portfd = port_create()) == -1) {
-    uv__set_sys_error(loop, errno);
-    return -1;
+  if(loop->fs_fd == -1) {
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: Style issue: space between if and paren.

File: src/unix/sunos.c [hunk 6]
```diff
@@ -144,39 +154,45 @@ int uv_fs_event_init(uv_loop_t* loop,
...
+    if ((portfd = port_create()) == -1) {
+      uv__set_sys_error(loop, errno);
+      return -1;
+    }
+    loop->fs_fd = portfd;
```
Comment (bnoordhuis) [inline src/unix/sunos.c:168]: The port fd is never closed. This looks like it needs a check in `uv_loop_delete()`.

**Comment by bnoordhuis** [comment]
Ira, can you sign [the CLA](http://nodejs.org/cla.html)?

File: src/unix/sunos.c [hunk 7]
```diff
@@ -128,13 +138,13 @@ static void uv__fs_event_read(EV_P_ ev_io* w, int revents) {
...
+    handle->fd = PORT_FIRED;
     handle->cb(handle, NULL, events, 0);
   }
-  while (handle->fd != -1);
+  while (handle->fd != PORT_DELETED);
 
-  if (handle->fd != -1)
-    uv__fs_event_rearm(handle);
+  if (handle->fd != PORT_DELETED) 
+    uv__fs_event_rearm(loop, handle);
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: loop_, not loop.

File: src/unix/sunos.c [hunk 8]
```diff
@@ -89,37 +94,42 @@ void uv_loadavg(double avg[3]) {
 
 
 #if HAVE_PORTS_FS
-static void uv__fs_event_rearm(uv_fs_event_t *handle) {
-  if (port_associate(handle->fd,
+static void uv__fs_event_rearm(uv_loop_t *loop, uv_fs_event_t *handle) {
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: You don't have to change the function prototype, `handle->loop` is a valid pointer.

File: src/unix/sunos.c [hunk 9]
```diff
@@ -144,39 +154,45 @@ int uv_fs_event_init(uv_loop_t* loop,
...
 }
 
 
 void uv__fs_event_destroy(uv_fs_event_t* handle) {
-  ev_ref(handle->loop->ev);
-  ev_io_stop(handle->loop->ev, &handle->event_watcher);
-  uv__close(handle->fd);
-  handle->fd = -1;
+  if(handle->fd == PORT_FIRED) {
+    port_dissociate(handle->loop->fs_fd, PORT_SOURCE_FILE, &handle->fo);
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: Needs a cast to `uintptr_t`

**Comment by ilc** [comment]
Contributor's agreement signed off.  Comments are dealt with.

File: include/uv.h [hunk 10]
```diff
@@ -1406,6 +1406,7 @@ struct uv_counters_s {
 
 struct uv_loop_s {
   UV_LOOP_PRIVATE_FIELDS
+  UV_LOOP_PRIVATE_FS_FIELDS
```
Comment (bnoordhuis) [inline include/uv.h:1409]: This breaks the Windows build.

**Comment by bnoordhuis** [comment]
@ilc There are still three or four unaddressed issues.

**Comment by ilc** [comment]
I think I found them all.  If there is anything left, please just call it out.

**Comment by bnoordhuis** [comment]
Thanks Ira. Merged in v0.6 in de64959 and browbeaten^Wmerged into master in 6865464.

**Comment by bnoordhuis** [comment]
Make that 7326962 and b5f9bc8 respectively, the old commits were breaking the build on other Unices.

**Pull request closed without merging**